### PR TITLE
docs: document ape.contracts public classes and methods

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

Document the contract module.

### How I did it

Document the two classes that users will be using from this module.
Make sure it all makes sense when building and looking at the docs.

### How to verify it

Builds the docs.
Make sure it makes sense.

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: src/ape/contracts/base.py [hunk 1]
```diff
@@ -185,6 +185,19 @@ class ContractEvent:
 
 
 class ContractInstance(AddressAPI):
+    """
+    An instance of a smart contract that you can interact with.
+    After you deploy a contract using the :class:`~ape.api.accounts.AccountAPI.deploy` method,
+    you get back a contract instance.
+
+    Usage example::
```
Comment (Ninjagod1251) [inline src/ape/contracts/base.py:193]: why do we :: ?

File: src/ape/contracts/base.py [hunk 2]
```diff
@@ -250,6 +300,18 @@ def __repr__(self) -> str:
         return f"<{self.contract_type.contractName}>"
 
     def at(self, address: str) -> ContractInstance:
+        """
+        Get a contract at the given address.
+
```
Comment (Ninjagod1251) [inline src/ape/contracts/base.py:305]: would a usage:
contract.at(add)?

File: src/ape/contracts/base.py (hunk 1 above)
Comment (unparalleled-js) [reply src/ape/contracts/base.py:193]: That is required to make it show up as the code-block style.

File: src/ape/contracts/base.py [hunk 3]
```diff
@@ -185,6 +185,19 @@ class ContractEvent:
 
 
 class ContractInstance(AddressAPI):
+    """
+    An instance of a smart contract that you can interact with.
```
Comment (evanblank3) [inline src/ape/contracts/base.py:189]:
```suggestion
    An interactive instance of a smart contract.
```

File: src/ape/contracts/base.py [hunk 4]
```diff
@@ -193,15 +206,40 @@ def __repr__(self) -> str:
...
             abi.name for abi in self._contract_type.abi
         ]
 
     def __getattr__(self, attr_name: str) -> Any:
+        """
+        Access a method or property on the contract using ``.`` access.
+
+        Usage example::
+
+            result = contract.Vote()  # Implies a method named Vote exists on the contract.
```
Comment (fubuloubu) [inline src/ape/contracts/base.py:234]: Methods are usually "camelCase"
```suggestion
            result = contract.vote()  # Implies a method named "vote" exists on the contract.
```

File: src/ape/contracts/base.py [hunk 5]
```diff
@@ -193,15 +206,40 @@ def __repr__(self) -> str:
...
+
+        Args:
+            attr_name (str): The name of the method or property to access.
+
+        Returns:
+            any: The return value from the contract call / transaction.
```
Comment (fubuloubu) [inline src/ape/contracts/base.py:240]:
```suggestion
            any: The return value from the contract call, or a transaction receipt.
```

**Review by fubuloubu — APPROVED** [review]
otherwise this lgtm

File: src/ape/contracts/base.py [hunk 6]
```diff
@@ -185,6 +185,19 @@ class ContractEvent:
...
+    An instance of a smart contract that you can interact with.
+    After you deploy a contract using the :class:`~ape.api.accounts.AccountAPI.deploy` method,
+    you get back a contract instance.
+
+    Usage example::
+
+        from ape import accounts, project
+
+        a = accounts.load("alias")  # Load an account by alias
+        contract = a.deploy(project.MyContract)  # The result of the deploy is a ContractInstance
```
Comment (evanblank3) [inline src/ape/contracts/base.py:198]:
```suggestion
        contract = a.deploy(project.MyContract)  # The result of ``deploy`` is a ContractInstance
```

File: src/ape/contracts/base.py [hunk 7]
```diff
@@ -250,6 +300,24 @@ def __repr__(self) -> str:
...
+        Get a contract at the given address.
+
+        Usage example::
+
+            from ape import project
+
+            my_contract = project.MyContract.at("0xAbC1230001112223334445566611855443322111")
+
+        Args:
+            address (str): The address to initialize the contract from.
```
Comment (evanblank3) [inline src/ape/contracts/base.py:313]:
```suggestion
            address (str): The address to initialize a contract.
```

File: src/ape/contracts/base.py [hunk 8]
```diff
@@ -250,6 +300,24 @@ def __repr__(self) -> str:
...
+              **NOTE**: Things will not work as expected if the contract is not actually
+              deployed to this address.
```
Comment (evanblank3) [inline src/ape/contracts/base.py:315]: Maybe also a note about if the contract type isn't what is expected as well?

File: src/ape/contracts/base.py (hunk 6 above)
Comment (unparalleled-js) [reply src/ape/contracts/base.py:198]: Since this is inside a code block, the tick marks don't get rendered

Comment (unparalleled-js) [reply src/ape/contracts/base.py:198]: Screen Shot 2021-12-20 at 15 32 28

Comment (unparalleled-js) [reply src/ape/contracts/base.py:198]: Screen Shot 2021-12-20 at 21 22 04

File: src/ape/contracts/base.py [hunk 9]
```diff
@@ -312,7 +314,8 @@ def at(self, address: str) -> ContractInstance:
         Args:
             address (str): The address to initialize a contract.
               **NOTE**: Things will not work as expected if the contract is not actually
-              deployed to this address.
+              deployed to this address or if the contract at the given address has
+              a different type than :attr:`~ape.contracts.ContractContainer.contract_type`.
```
Comment (fubuloubu) [inline src/ape/contracts/base.py:318]:
```suggestion
              a different ABI than :attr:`~ape.contracts.ContractContainer.contract_type`.
```

File: src/ape/contracts/base.py (hunk 6 above)
Comment (unparalleled-js) [reply src/ape/contracts/base.py:198]: hmm, maybe something like this then:

```
# The result of 'deploy()' is a ContractInstance
```

**Review by evanblank3 — APPROVED** [review]

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**